src/main.go

```diff
@@ -284,9 +334,114 @@ func garbageClean(directory, cname, current_version string, size_wanted int64) e
 	return nil
 }
 
-const ERR_INVALID_ARGUMENTS = 1
-const ERR_SYSTEM_FAILURE = 2
-const ERR_NETWORK_PROBLEMS = 3
+func verifyManifest(repo *remote.Repository, ctx context.Context, digest, verificationKeyFile string) {
+	signatureTag := strings.Replace(digest, "sha256:", "sha256-", 1) + ".sig"
+	signatureManifestBytes, err := getManifestBytes(repo, ctx, signatureTag)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error:", err)
+		os.Exit(ERR_NETWORK_PROBLEMS)
+	}
+
+	signatureManifest := SignatureManifest{}
+	err = json.Unmarshal(signatureManifestBytes, &signatureManifest)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error:", err)
+		os.Exit(ERR_INVALID_FORMAT)
+	}
+	// types
+	signatureStr := signatureManifest.Layers[0].Annotations.Signature
+	signature, err := base64.StdEncoding.DecodeString(signatureStr)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error:", err)
+		os.Exit(ERR_INVALID_FORMAT)
+	}
+
+	messageHashStr := signatureManifest.Layers[0].Digest
+	messageHashFromManifest, err := hex.DecodeString(strings.Trim(messageHashStr, "sha256:"))
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error:", err)
+		os.Exit(ERR_INVALID_FORMAT)
+	}
+
+	// Here we pull the messageHashStr. This is insufficient for a proper signature verification. We have to
+	// check the messages contents, validate that it contains the correct manifest digest, and then hash it ourselves.
```

Comment:
looks a bit needlessly complicated to me, but fine, I guess it has to do with how cosign works.... simply building the message out of the digest doesn't work?

Reply:
Yes, building the message locally would work too. 
The problem I see is that there may be multiple valid JSON-objects that contain the digest. But even if we assume that there are is only one valid JSON-object there are multiple valid messages to represent that JSON-object (due to indentation for example). Therefore we chose the approach to load the original message, load it into a JSON object and then validate that the digest is correct. I think OCI registries normalize JSON strings so that there are reproducible digests for the JSON-objects. However this only works for manifests, this message is stored as a blob.